File: backend/shared/validation.py
"""Shared validation constants for LC_Care backend."""

# ── Enum value sets ──
GENDERS = {"male", "female"}
CARE_FOR = {"self", "child", "parents", "family"}
HEALTH_STATUS = {"healthy", "monitoring", "chronic", "mental", "pregnant", "recovery"}
CHRONIC_SUB_CONDITIONS = {"diabetes", "hypertension", "cholesterol", "bone_joint", "digestive"}
EXERCISE = {"active", "sometimes", "sedentary", "starting"}
BAD_HABITS = {"latenight", "alcohol", "unhealthy_eating", "stress", "smoking", "skip_meals", "sleep_late", "ok"}
LIFE_CHANGES = {"job", "move", "pregnant", "baby", "retired", "travel", "none"}
PLANT_STATUS = {"pending", "growing", "paused", "graduated"}
JOURNEY_TYPES = {
    "health_care", "monitoring", "prevention", "screening", "maternity",
    "pediatric", "long_term_adherence", "mental_wellness", "recovery",
    "digestive_health", "bone_health",
}
RELATIONS = {"self", "spouse", "child", "parent", "sibling", "grandparent", "other"}

# ── Field length/int limits ──
NAME_MAX_LENGTH = 100
NOTE_MAX_LENGTH = 2000
YEAR_MIN = 1900
YEAR_MAX = 2016
AGE_MIN = 0
AGE_MAX = 150
CYCLE_LENGTH_MIN = 18
CYCLE_LENGTH_MAX = 45
PLANT_LEVEL_MIN = 1
PLANT_LEVEL_MAX = 10
STREAK_DAYS_MAX = 3650
HABIT_ID_MIN = 0
CONFIDENCE_MIN = 0.0
CONFIDENCE_MAX = 1.0

# ── Collection limits ──
PURCHASE_ITEMS_MIN = 1
FAMILY_MEMBERS_MIN = 1
FAMILY_MEMBERS_MAX = 6
CROSS_MEMBER_MAX = 20

# ── Vietnamese error messages ──
ERRORS = {
    "required": "Vui lòng nhập {field}",
    "invalid_enum": "{field} không hợp lệ. Các giá trị chấp nhận: {values}",
    "min_length": "{field} phải có ít nhất {min} ký tự",
    "max_length": "{field} không được vượt quá {max} ký tự",
    "year_range": "Năm sinh phải từ {min} đến {max}",
    "age_range": "Tuổi phải từ {min} đến {max}",
    "ge": "{field} phải lớn hơn hoặc bằng {min}",
    "le": "{field} phải nhỏ hơn hoặc bằng {max}",
    "min_items": "Phải có ít nhất {min} mục",
    "max_items": "Không được vượt quá {max} mục",
}

File: frontend/src/utils/validation.js
// Validation helpers for LC_Care onboarding forms
// Mirrors backend/shared/validation.py constants
import i18n from "../i18n";
const t = (k, o) => i18n.t(k, o);

// ── Enum value sets ──
const GENDERS = ["male", "female"];
const CARE_FOR = ["self", "child", "parents", "family"];
const HEALTH_STATUS = ["healthy", "monitoring", "chronic", "mental", "pregnant", "recovery"];
const CHRONIC_SUB_CONDITIONS = ["diabetes", "hypertension", "cholesterol", "bone_joint", "digestive"];
const EXERCISE = ["active", "sometimes", "sedentary", "starting"];
const BAD_HABITS = ["latenight", "alcohol", "unhealthy_eating", "stress", "smoking", "skip_meals", "sleep_late", "ok"];
const LIFE_CHANGES = ["job", "move", "pregnant", "baby", "retired", "travel", "none"];

// ── Field limits ──
const NAME_MAX_LENGTH = 100;
const YEAR_MIN = 1900;
const YEAR_MAX = 2016;
const CYCLE_LENGTH_MIN = 18;
const CYCLE_LENGTH_MAX = 45;
const NOTE_MAX_LENGTH = 2000;

// ── Welcome Step ──
export function validateWelcomeStep(data) {
  const errors = {};
  const name = (data.fullName || "").trim();

  if (!name) {
    errors.fullName = t("validation_name_required");
  } else if (name.length > NAME_MAX_LENGTH) {
    errors.fullName = t("validation_name_length", { max: NAME_MAX_LENGTH });
  }

  if (!GENDERS.includes(data.gender)) {
    errors.gender = t("validation_gender_required");
  }

  const year = parseInt(data.yearOfBirth, 10);
  if (isNaN(year) || year < YEAR_MIN || year > YEAR_MAX) {
    errors.yearOfBirth = t("validation_birth_year_range", { min: YEAR_MIN, max: YEAR_MAX });
  }

  return errors;
}

// ── Health Scan Step ──
export function validateHealthScanStep(data) {
  const errors = {};

  if (!HEALTH_STATUS.includes(data.healthStatus)) {
    errors.healthStatus = t("validation_health_status");
  }

  if (data.healthStatus === "chronic" && (!data.chronicSubConditions || data.chronicSubConditions.length === 0)) {
    errors.chronicSubConditions = t("validation_chronic");
  }

  return errors;
}

// ── Cycle Tracking Step ──
export function validateCycleTrackingStep(data) {
  const errors = {};

  if (data.lastDate) {
    const d = new Date(data.lastDate);
    if (isNaN(d.getTime())) {
      errors.lastPeriodDate = t("validation_date_invalid");
    } else if (d > new Date()) {
      errors.lastPeriodDate = t("validation_date_future");
    }
  }

  const cycleLen = parseInt(data.cycleLength, 10);
  if (isNaN(cycleLen) || cycleLen < CYCLE_LENGTH_MIN || cycleLen > CYCLE_LENGTH_MAX) {
    errors.cycleLength = t("validation_cycle_length", { min: CYCLE_LENGTH_MIN, max: CYCLE_LENGTH_MAX });
  }

  return errors;
}

// ── Habits Step ──
export function validateHabitsStep(data) {
  const errors = {};

  if (!EXERCISE.includes(data.exercise)) {
    errors.exercise = t("validation_exercise");
  }

  return errors;
}

// ── Pharmacist Note ──
export function validatePharmacistNote(note) {
  const errors = {};
  if (note && note.length > NOTE_MAX_LENGTH) {
    errors.note = t("validation_note_length", { max: NOTE_MAX_LENGTH });
  }
  return errors;
}
